pyensign/ensign.py

```diff
@@ -818,17 +819,19 @@ async def wrapper(*args, **kwargs):
     # Return either a coroutine or an async generator wrapper to match the marked
     # function
     def decorator(
-        coro: Union[Coroutine[Any, Any, Any], AsyncGenerator[Any, None]]
+        fn: Union[Coroutine[Any, Any, Any], AsyncGenerator[Any, None]]
     ) -> Union[Coroutine[Any, Any, Any], AsyncGenerator[Any, None]]:
-        if inspect.iscoroutinefunction(coro):
-            return wrap_coroutine(coro)
-        elif inspect.isasyncgenfunction(coro):
-            return wrap_async_generator(coro)
+        if inspect.isgeneratorfunction(fn):
+            return async_to_sync(wrap_async_generator(sync_to_async(fn)))
+        elif inspect.isasyncgenfunction(fn):
+            return wrap_async_generator(fn)
+        elif inspect.iscoroutinefunction(fn):
+            return wrap_coroutine(fn)
+        elif inspect.isfunction(fn):
+            return async_to_sync(wrap_coroutine(sync_to_async(fn)))
```

Comment:
I think some commenting inside of this function and a docstring rather that overhead # comments would be very helpful. I'm having a hard time understanding why we're making the `fn` aync, wrapping it, then making it `sync` without any context.